src/test/java/org/mars/rover/kata/StdinProcessorTest.java

```diff
@@ -0,0 +1,39 @@
+package org.mars.rover.kata;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mars.rover.kata.commands.CommandParser;
+import org.mars.rover.kata.entrydata.StdinProcessor;
+
+import java.io.ByteArrayInputStream;
+import java.util.Scanner;
+
+public class StdinProcessorTest {
+
+    @Test
+    public void receivesStringInputAndConstructsGridDimensions() {
+        // given
+        String coordinatesString = "5 5\n" + "1 2 N\n" + "LMLMLMLMM\n" + "3 3 E\n" + "MMRMMRMRRM\n";
+        System.setIn(new ByteArrayInputStream(coordinatesString.getBytes()));
+        var stringInputProcessor = new StdinProcessor(new CommandParser());
+
+        var commandSet = stringInputProcessor.processInput(new Scanner(System.in));
+
+        // expect
+        Assertions.assertEquals(5, commandSet.gridX());
+        Assertions.assertEquals(5, commandSet.gridY());
+        Assertions.assertEquals(2, commandSet.roverInstructions().size());
+    }
+
+    @Test
+    public void transformsStringInputToASetOfCommands() {
+        // given
+        String coordinatesString = "5 5\n" + "1 2 N\n" + "LMLMLMLMM\n" + "3 3 E\n" + "MMRMMRMRRM\n";
+        System.setIn(new ByteArrayInputStream(coordinatesString.getBytes()));
+        var stringInputProcessor = new StdinProcessor(new CommandParser());
+        var commandSet = stringInputProcessor.processInput(new Scanner(System.in));
```

Comment:
OK, patrząc na to wywołanie myślę warto poruszyć dwie powszechne szkoły architektoniczne.

1. Gloryfikujące [anemiczne obiekty](https://en.wikipedia.org/wiki/Anemic_domain_model) (tylko dane)
2. Gloryfikujące [samo-zarządzające sobą obiekty](https://www.baeldung.com/java-anemic-vs-rich-domain-objects) (dane + metody)

Ty poszedłeś drogą #1, tzn. obiekty zawierają dane, a logika wykonuje się w osobnych klasach (`CośtamProcessor`, `Cośtam(Application)Service`, itd...). Jest to droga popularyzowana przez purystów Object Oriented Programming, trochę też w Clean Code. W tej drodze pojawia się dużo interfejsów, dużo klas (choćby jedna na dane, druga na operacje na nich, ale zwykle więcej). To tutaj pojawiają się `CośtamCommandInterface` i `AbstractMethodFactoryInterface` itd. Nie ma nic w tym złego i jest to ostateczna forma [zasady hermetyzacji](https://pl.wikipedia.org/wiki/Hermetyzacja_(informatyka)) w obiektach i [prawa demeter](https://pl.wikipedia.org/wiki/Prawo_Demeter).

Moim zdaniem to podejście nie nadaje się do większości projektów, jakie realnie piszemy, tzn. małych do średnich, gdzie backend bezpośrednio wykonuje działania na życzenie frontendu. Sprawdza się za to przy pisaniu systemów, lub częściej bibliotek, gdzie dodatkowy nakład włożony w abstrachowanie zwróci się.

Dlatego do "naszych typowych" aplikacji wolę [Domain-Driven Design](https://pl.wikipedia.org/wiki/Domain-driven_design). Spójrzmy na to pragmatycznie: czy w tak małym projekcie, na tym etapie zupełnie początkowym, przewidujemy w ogóle istnienie komend czy grida, który tworzony inną metodą niż z tego jednego określonego formatu stringowego?

Dlatego ja osobiście poszedłbym w odwrotnym kierunku. Pseudo kod dla zilustrowania:

```java
  var lineScanner = new Scanner(System.in);

  var gridSize = lineScanner.nextLine();
  var grid = Grid.fromInput(gridSize);

  var initRoverPosition = lineScanner.nextLine();
  var rover = MarsRover.fromInput(initRoverPosition);

  var commandParser = new CommandParser();
  var commands = lineScanner.nextLine();
  rover = commands.chars()
    .mapToObj(i -> (char)i) // Bo Java zdecydowała zwrócić IntStream z `chars()`
    .map(commandParser::parse)
    .reduce(() -> rover, (rover, command) -> rover.executeCommand(command));
```

W ten sposób zarówno `Grid` jak i `MarsRover` potrafią się same stworzyć z odpowiedniego inputu i to one obsługują błędy. Czyli mamy *rich objects*, które prócz danych potrafią się same obsłużyć.

To jest głębszy temat i myślę, że powinniśmy go omówić na następnym spotkaniu.